### PR TITLE
[Feature][Model] Support MiniMax M3 MXFP8 inference on Ascend 950

### PR DESCRIPTION
**Pull request by HaoxinZong**
### What this PR does / why we need it?

This PR adds MiniMax M3 MXFP8 inference support for Ascend A5 while preserving the existing Ascend A3 execution paths.

- Add an A5-specific sparse-attention prefill path based on K2Q CSR and `npu_sparse_attention_score_prefill`, while retaining the existing A3 sparse-attention interface.
- Dispatch MiniMax M3 index score, top-k, and decode to the A5 Triton implementations on A5, without changing the A3 implementations.
- Add the required custom-op sources, Torch bindings, metadata, build selection, and operator tests for the A5 path.
- Integrate ModelOpt-compatible MXFP8 checkpoints with the Ascend linear and fused-MoE quantization methods.
- Select `scale_alg=1` only for MiniMax M3 on Ascend A5 and keep the default algorithm for other devices and models.
- Keep embedding and LM head layers unquantized according to the checkpoint configuration, and keep the KV cache in BF16.
- Use the device-op abstraction for clipped SwiGLU and retain the small-operator QKV/RMSNorm/RoPE path required by A5.
- Align the implementation with the vLLM Ascend main branch and the vLLM 0.27.1 compatibility baseline.

Ascend A3 and A5 expose different sparse-attention and cache-update capabilities. The device-specific dispatch introduced here keeps each platform on its supported operator interface without changing the compatible framework-level behavior.

### Does this PR introduce _any_ user-facing change?

Yes. MiniMax M3 models with ModelOpt-compatible MXFP8 weights can run on Ascend A5 with BF16 KV cache, while the existing Ascend A3 path remains unchanged.

### How was this patch tested?

- `git diff --check`
- Python syntax compilation for the changed Python modules
- `bash -n csrc/build_aclnn.sh`
- Audited A3/A5 sparse-attention, indexer, and decode dispatch paths
- Added unit coverage for A3/A5 sparse-attention dispatch, A5 index decode metadata, ModelOpt MXFP8 configuration, and MX quant scale-algorithm selection

- vLLM version: v0.27.1
- vLLM main: https://github.com/vllm-project/vllm/commit/58d3918e3ea0a544ffedadad2ba84559e9c51d8f

### Co-authors

- **realliujiaxu** (<realliujiaxu@163.com>) — Solution design and Triton kernel optimization
- **wlf-darkmatter** (<wlf_x_contion@sina.com>) — K2Q CSR operator implementation
- **gxy9808** (<gxywhut@163.com>) — Accuracy and performance evaluation, operator functional testing
- **NengXu001** (<xuneng1998@gmail.com>) — Accuracy and performance testing

### CODE REVIEW

File: csrc/attention/k2q_csr/k2q_csr_common/op_host/arch35/k2q_csr_tiling_arch35.cpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ *
```
Comment (chenlongxiao) [inline csrc/attention/k2q_csr/k2q_csr_common/op_host/arch35/k2q_csr_tiling_arch35.cpp:2]: 补充MD

File: vllm_ascend/models/minimax_m3/minimax_m3.py [hunk 2]
```diff
@@ -96,6 +98,46 @@
...
+        key_cache = cache.unsqueeze(2)
+        torch_npu.npu_scatter_pa_cache(
+            key,
+            slots.contiguous(),
+            key_cache=key_cache,
+        )
+        return
+
+    flat_cache = cache.view(-1, cache.shape[-1])
+    torch.ops._C_ascend.npu_scatter_nd_update_v2(
```
Comment (weijinqian0) [inline vllm_ascend/models/minimax_m3/minimax_m3.py:131]: Check whether the npu_scatter_nd_update_v2 function is required.

File: vllm_ascend/ops/fused_moe/router/fused_topk_router.py [hunk 3]
```diff
@@ -142,7 +143,11 @@ def _compute_routing(
             )
             return topk_weights, topk_ids
         norm_type = 0 if self.scoring_func == "softmax" else 1
-        if self.e_score_correction_bias is not None and self.e_score_correction_bias.dtype != router_logits.dtype:
+        if (
+            self.e_score_correction_bias is not None
+            and self.e_score_correction_bias.dtype != router_logits.dtype
+            and not self.preserve_e_score_correction_bias_dtype
+        ):
             self.e_score_correction_bias = self.e_score_correction_bias.to(router_logits.dtype)
```
Comment (weijinqian0) [inline vllm_ascend/ops/fused_moe/router/fused_topk_router.py:146]: This modification is not required. router_logits is of the fp32 type.

File: vllm_ascend/ops/fused_moe/moe_mlp.py [hunk 4]
```diff
@@ -112,11 +113,62 @@ def _prepare_dequant_swiglu_weight_scale(
...
+
+
+def _apply_clipped_swiglu(
+    hidden_states: torch.Tensor,
+    *,
+    swiglu_limit: float,
+    swiglu_alpha: float,
+    swiglu_beta: float,
+) -> torch.Tensor:
+    if ASCEND_DEVICE_TYPE == AscendDeviceType.A5:
```
Comment (weijinqian0) [inline vllm_ascend/ops/fused_moe/moe_mlp.py]: You can adapt the hardware using device_op.

File: vllm_ascend/patch/platform/patch_fused_moe.py [hunk 5]
```diff
@@ -166,6 +193,13 @@ def _ascend_FusedMoE(
         routed_experts_args=routed_experts_args,
         **kwargs,
     )
+    if hasattr(runner.router, "preserve_e_score_correction_bias_dtype"):
+        runner.router.preserve_e_score_correction_bias_dtype = (
+            _preserve_routing_bias_fp32(
+                get_current_vllm_config(),
+                runner.routed_experts.quant_method,
+            )
+        )
```
Comment (weijinqian0) [inline vllm_ascend/patch/platform/patch_fused_moe.py]: plz remove this.

File: vllm_ascend/quantization/utils.py [hunk 6]
```diff
@@ -37,6 +37,28 @@
...
+    if isinstance(architectures, str):
+        architectures = (architectures,)
+    is_minimax_m3 = any(
+        isinstance(architecture, str) and architecture.startswith("MiniMaxM3") for architecture in architectures
+    )
+    hf_text_config = getattr(model_config, "hf_text_config", None)
+    is_minimax_m3 = is_minimax_m3 or getattr(hf_text_config, "model_type", None) == "minimax_m3"
+    return int(is_minimax_m3)
+
+
```
Comment (weijinqian0) [inline vllm_ascend/quantization/utils.py:69]: You can add an lru_cache annotation here.

File: vllm_ascend/patch/platform/patch_fused_moe.py [hunk 7]
```diff
@@ -40,13 +40,40 @@
...
+    is_minimax_m3 = is_minimax_m3 or (
+        getattr(hf_text_config, "model_type", None) == "minimax_m3"
+    )
+
+    quant_scheme = getattr(quant_method, "quant_method", quant_method)
+    is_mxfp8 = (
+        quant_scheme.__class__.__name__
+        == "AscendW8A8MXFP8DynamicFusedMoEMethod"
+    )
+    return is_minimax_m3 and is_mxfp8
```
Comment (weijinqian0) [inline vllm_ascend/patch/platform/patch_fused_moe.py]: Plz  remove this.

File: vllm_ascend/quantization/modelslim_config.py [hunk 8]
```diff
@@ -592,8 +624,17 @@ def from_config(cls, config: dict[str, Any]) -> "AscendModelSlimConfig":
     def override_quantization_method(cls, hf_quant_cfg, user_quant, hf_config: Any = None) -> str | None:
         if hf_quant_cfg is not None:
             quant_method = hf_quant_cfg.get("quant_method", None)
-            if not quant_method and torch.npu.is_available():
-                return ASCEND_QUANTIZATION_METHOD
+            if torch.npu.is_available():
+                if not quant_method:
+                    return ASCEND_QUANTIZATION_METHOD
+                if quant_method == "mxfp8" and get_ascend_device_type() == AscendDeviceType.A5:
```
Comment (weijinqian0) [inline vllm_ascend/quantization/modelslim_config.py]: Do not add hardware judgment.

File: vllm_ascend/quantization/utils.py [hunk 9]
```diff
@@ -37,6 +38,43 @@
...
+        return 0
+    if any(architecture.startswith("MiniMaxM3") for architecture in architectures) or model_type == "minimax_m3":
+        return 1
+    return 0
+
+
+def get_dynamic_mx_quant_scale_alg(vllm_config=None) -> int:
+    """Select the MiniMax M3-specific MX quant scale algorithm."""
+    device_type = get_ascend_device_type()
+    if device_type != AscendDeviceType.A5:
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/utils.py]: The device type has been determined here, which overlaps with the logic in `_get_dynamic_mx_quant_scale_alg`. Therefore, `_get_dynamic_mx_quant_scale_alg` is not necessary and only `get_dynamic_mx_quant_scale_alg` is reserved.

File: vllm_ascend/quantization/modelslim_config.py [hunk 10]
```diff
@@ -553,12 +553,35 @@ class AscendModelSlimConfig(QuantizationConfig):
     def __init__(self, quant_config: dict[str, Any] | None = None):
         super().__init__()
         self.quant_description = quant_config if quant_config is not None else {}
+        self._is_mxfp8_checkpoint = False
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/modelslim_config.py]: The community has a corresponding quantization method for minimax-mxfp8 weights called `ModelOptMxFp8Config`. It should refer to DeepSeek V4, inherit the community's quantization approach, and rewrite the methods we need, rather than making invasive modifications to `modelslim_config`.

File: vllm_ascend/ops/fused_moe/moe_mlp.py [hunk 11]
```diff
@@ -243,44 +295,71 @@ def quant_apply_mlp(
...
-                        "glu_alpha": swiglu_alpha,
-                        "glu_bias": swiglu_beta,
-                    }
-                )
-            hidden_states, swiglu_out_scale = torch.ops._C_ascend.npu_dequant_swiglu_quant(**dequant_swiglu_kwargs)
+            hidden_states, swiglu_out_scale = DeviceOperator.npu_dynamic_quant(
+                hidden_states, act_quant_type=act_quant_type, use_mxfp_quant=use_mxfp_quant
+            )
+        else:
+            if use_mxfp_quant and is_swigluoai_uninterleave:
```
Comment (kunpengW-code) [inline vllm_ascend/ops/fused_moe/moe_mlp.py]: The moe_mlp.py has been reconstructed and eliminated in the PR https://github.com/vllm-project/vllm-ascend/pull/13773. It is recommended that the reconstruction be incorporated and then new modifications be performed.

File: vllm_ascend/quantization/utils.py [hunk 12]
```diff
@@ -37,6 +38,43 @@
...
+@lru_cache(maxsize=None)
+def _get_dynamic_mx_quant_scale_alg(
+    device_type: AscendDeviceType,
+    architectures: tuple[str, ...],
+    model_type: str | None,
+) -> int:
+    if device_type != AscendDeviceType.A5:
+        return 0
+    if any(architecture.startswith("MiniMaxM3") for architecture in architectures) or model_type == "minimax_m3":
+        return 1
```
Comment (Biuapha) [inline vllm_ascend/quantization/utils.py]: magic number？It's not easy to understand

File: vllm_ascend/models/minimax_m3/ops/msa_m3_triton_a5.py [hunk 13]
```diff
@@ -0,0 +1,1454 @@
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright contributors to the vLLM project
+"""Triton kernels for MiniMax M3 block-sparse GQA attention on Ascend.
+
+Migrated from reference/vllm_cp/vllm/models/minimax_m3/common/ops/sparse_attn.py.
```
Comment (Biuapha) [inline vllm_ascend/models/minimax_m3/ops/msa_m3_triton_a5.py:5]: a3,a5 should use the same triton kernel

File: vllm_ascend/quantization/utils.py [hunk 14]
```diff
@@ -37,6 +38,43 @@
...
 
+@lru_cache(maxsize=None)
+def _get_dynamic_mx_quant_scale_alg(
+    device_type: AscendDeviceType,
+    architectures: tuple[str, ...],
+    model_type: str | None,
+) -> int:
+    if device_type != AscendDeviceType.A5:
+        return 0
+    if any(architecture.startswith("MiniMaxM3") for architecture in architectures) or model_type == "minimax_m3":
```
Comment (Biuapha) [inline vllm_ascend/quantization/utils.py]: use model architecture is bad，quant config is better

File: vllm_ascend/models/minimax_m3/ops/msa_m3_triton_a5.py (hunk 13 above)
Comment (HaoxinZong) [reply vllm_ascend/models/minimax_m3/ops/msa_m3_triton_a5.py:5]: A TODO has been added ： Unify the A3 and A5 sparse-attention kernels once their operator
# interfaces and implementations are compatible. They must remain separate
# for now.

File: vllm_ascend/quantization/modelopt_mxfp8_config.py [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+    def mxfp8_moe_weight_loader(
+        param: torch.nn.Parameter,
+        loaded_weight: torch.Tensor,
+        weight_name: str,
+        shard_id: str,
+        expert_id: int,
+        return_success: bool = False,
+    ) -> bool | None:
+        prefix, separator, param_name = weight_name.rpartition(".")
+        mapped_name = _MOE_WEIGHT_LOADER_NAME_MAP.get(param_name, param_name)
```
Comment (kunpengW-code) [inline vllm_ascend/quantization/modelopt_mxfp8_config.py]: What is the reason for applying the patch here and handling `_MOE_WEIGHT_LOADER_NAME_MAP`? I checked the original weights, and there are no parameters like `scale_bias` in the weights.

Comment (HaoxinZong) [reply vllm_ascend/quantization/modelopt_mxfp8_config.py]: I ve checked and comfirmed this modification is unnesscesry and removed it

**Review by Pz1116 — APPROVED** [review]
LGTM, need another approve from kernal maintainers and also Model feature owner

File: csrc/attention/k2q_csr/k2q_csr_common/op_host/arch35/k2q_csr_tiling_arch35.cpp (hunk 1 above)
Comment (chenlongxiao) [inline csrc/attention/k2q_csr/k2q_csr_common/op_host/arch35/k2q_csr_tiling_arch35.cpp:2]: 补充MD

**Review by chenlongxiao — APPROVED** [review]

**Review by chenlongxiao — APPROVED** [review]

**Review by realliujiaxu — DISMISSED** [review]

**Review by realliujiaxu — APPROVED** [review]

**Review by zzzzwwjj — APPROVED** [review]
